File: app/handlers/video.py
import asyncio
import logging
import random
import tempfile

from aiogram import Bot, F, Router, exceptions
from aiogram.types import FSInputFile, Message
from aiogram.utils import chat_action

from app.services.video import download_video
from app.settings import settings

logger = logging.getLogger(__name__)

router = Router(name=__name__)


@router.message(F.text.startswith('https://x.com'))
async def handle_link_handler(message: Message, bot: Bot) -> None:
    assert message.text is not None

    bot_instance = await bot.me()

    if bot_instance.username == settings.OLD_USERNAME:
        await message.reply(
            'Please, use @StarlightManagerBot instead. This bot is deprecated, thanks.'
        )

    async with chat_action.ChatActionSender.upload_video(
        chat_id=message.chat.id,
        bot=message.bot,  # pyright: ignore[reportArgumentType]
    ):
        logger.info('Trying to download %s', message.text)

        with tempfile.TemporaryDirectory() as tmpdir:
            logger.debug('Using temporary directory %s', tmpdir)

            try:
                videos = download_video(message.text, tmpdir)
            except Exception:
                logger.exception('Error downloading video')
                await message.reply(
                    'Error downloading video, probably X is blocking the request.',
                )
                return

            # TODO: Change to media group
            for video in videos:
                try:
                    await message.reply_video(
                        video=FSInputFile(video.file_path),
                        filename=video.file_path.name,
                        height=video.metadata.height,
                        width=video.metadata.width,
                    )
                    await asyncio.sleep(random.randint(1, 2))
                except exceptions.TelegramAPIError:
                    logger.exception('Telegram API error')
                    await message.reply(
                        'Error happened while uploading video, probably it is too large.',
                    )
                    return
                except FileNotFoundError:
                    logger.exception('File not found')
                    await message.reply('File not found, probably could not be downloaded.')
                    return


@router.message()
async def handle_message(message: Message) -> None:
    await message.reply('Please, provide a valid link to X post.')

File: backend/src/handlers/video.ts
import type { Context } from "@/bot";
import { downloadVideo } from "@/services/video";
import { Composer, GrammyError, InputFile } from "grammy";
import tmp from "tmp";

const composer = new Composer<Context>();

const feature = composer.chatType("private");

feature.on(":text").filter(
	(ctx) =>
		ctx.msg.text.startsWith("https://x.com") ||
		ctx.msg.text.startsWith("https://www.instagram.com") ||
		ctx.msg.text.startsWith("https://instagram.com"),
	async (ctx) => {
		await ctx.replyWithChatAction("upload_video");

		const tempDir = tmp.dirSync({ unsafeCleanup: true });
		const videos = await downloadVideo(ctx.msg.text, tempDir.name);

		for (const video of videos) {
			try {
				await ctx.replyWithVideo(new InputFile(video.filePath), {
					width: video.metadata?.width,
					height: video.metadata?.height,
				});
			} catch (error) {
				if (error instanceof GrammyError) {
					if (error.error_code === 413) {
						await ctx.reply("Video is too large, can't be sent");
					} else {
						await ctx.reply("Error downloading video");
						throw error;
					}
				}
			}
		}

		tempDir.removeCallback();
	},
);

export default composer;
